core/ffmpeg/transcoder.go

```diff
@@ -0,0 +1,285 @@
+		return widthString + ":-2"
+	} else if heightString != "0" {
+		return "-2:" + heightString
+	}
+
+	return ""
+}
+
+func (t *Transcoder) Start() error {
+	command := t.GetString()
+
+	_, err := exec.Command("sh", "-c", command).Output()
+	if err != nil {
+		log.Panicln(err, command)
+	}
+
+	log.Println("Video transcoder started.")
+
+	return err
+}
+
+func (t *Transcoder) GetString() string {
+	hlsOptionFlags := []string{
+		"delete_segments",
+		"program_date_time",
+		"temp_file",
+	}
+	ffmpegFlags := []string{
+		"cat", t.Input, "|",
+		config.Config.FFMpegPath,
+		"-hide_banner",
+		"-i pipe:",
+		t.getVariantsString(),
+
+		// HLS Output
+		"-f", "hls",
+		"-hls_time", strconv.Itoa(config.Config.VideoSettings.ChunkLengthInSeconds), // Length of each segment
+		"-hls_list_size", strconv.Itoa(config.Config.Files.MaxNumberInPlaylist), // Max # in variant playlist
+		"-hls_delete_threshold", "10", // Start deleting files after hls_list_size + 10
+		"-hls_flags", strings.Join(hlsOptionFlags, "+"), // Specific options in HLS generation
+
+		// Video settings
+		"-tune", "zerolatency", // Option used for good for fast encoding and low-latency streaming (always includes iframes in each segment)
+		// "-profile:v", "high", // Main – for standard definition (SD) to 640×480, High – for high definition (HD) to 1920×1080
+		"-sc_threshold", "0", // Disable scene change detection for creating segments
+
+		// Filenames
+		"-master_pl_name", "stream.m3u8",
+		"-strftime 1",                                                               // Support the use of strftime in filenames
+		"-hls_segment_filename", path.Join(t.SegmentOutputPath, "/%v/stream-%s.ts"), // Each segment's filename
+		"-max_muxing_queue_size", "9999", // Workaround for Too many packets error: https://stackoverflow.com/questions/49686244/ffmpeg-too-many-packets-buffered-for-output-stream-01
+		path.Join(t.SegmentOutputPath, "/%v/stream.m3u8"), // Each variant's playlist
+	}
+
+	return strings.Join(ffmpegFlags, " ")
+}
+
+func getVariantFromConfigQuality(quality config.StreamQuality, index int) HLSVariant {
+	variant := HLSVariant{}
+	variant.Index = index
+	variant.IsAudioPassthrough = quality.IsAudioPassthrough
+	variant.IsVideoPassthrough = quality.IsVideoPassthrough
+	variant.EncoderPreset = quality.EncoderPreset
+
+	// If no audio bitrate is specified then we pass through original audio
+	if quality.AudioBitrate == 0 {
+		variant.IsAudioPassthrough = true
+	}
+
+	if quality.Bitrate == 0 {
+		variant.IsVideoPassthrough = true
+	}
+
+	if variant.IsVideoPassthrough {
+		return variant
+	}
+
+	variant.SetVideoBitrate(strconv.Itoa(quality.Bitrate) + "k")
+	variant.SetAudioBitrate(strconv.Itoa(quality.AudioBitrate) + "k")
+	variant.SetVideoScalingWidth(quality.ScaledWidth)
+	variant.SetVideoScalingHeight(quality.ScaledHeight)
+	variant.SetVideoFramerate(quality.Framerate)
+
+	return variant
+}
+
+// CreateTranscoderFromConfig will return a new Transcoder, populated by the config
```

Comment:
I don't really like this implicit initialization via the config, instead of the config being explicitly passed in, but we're pulling the config elsewhere, so I thought I'd go that route, but I don't know if it should be kept.

Reply:
I think you could just name it `NewTranscoder` unless you plan to create another way that someone can create a valid Transcoder to be used. Unless I am missing something, we currently assume everything is done via the config file.